liquidity-pool-changes/src/agent.js

```diff
@@ -0,0 +1,219 @@
+      token1Address = await poolContract.token1();
+    } catch {
+      // asume its not a Uniswap V3 pool if contract calls fail
+      return undefined;
+    }
+
+    const token0Contract = new ethers.Contract(
+      token0Address,
+      erc20Abi,
+      provider,
+    );
+    const token1Contract = new ethers.Contract(
+      token1Address,
+      erc20Abi,
+      provider,
+    );
+
+    // get amount of eth in the pool if applicable (if the pool is using WETH, then this value is 0)
+    const poolEthBalance = await provider.getBalance(poolContract.address);
+
+    // get the # amount of each tokens in the liquidity pool
+    const tokenBalance0 = await token0Contract.balanceOf(poolContract.address);
+    const tokenBalance1 = await token1Contract.balanceOf(poolContract.address);
+
+    // convert # amount of each tokens in the liquidity pool from ethers BigNumber to BigNumber.js
+    const tokenBalance0BN = new BigNumber(tokenBalance0.toString());
+    const tokenBalance1BN = new BigNumber(tokenBalance1.toString());
+    const poolEthBalanceBN = new BigNumber(poolEthBalance.toString());
+
+    // this returns the token prices in usd of each token (without accounting for decimals)
+    let tokenPrices;
+    try {
+      tokenPrices = await getTokenPrices(token0Address, token1Address);
+    } catch (error) {
+      // if coingecko call fails
+      return findings;
+    }
+
+    // multiply the amount of tokens in the pool by usd value, adjusting for decimals
+    let token0ValueUSD;
+    let token1ValueUSD;
+    let currentLiquidity;
+
+    if (tokenBalance0BN.gt(0)) {
+      token0ValueUSD = await getTokenUSDValue(
+        tokenBalance0BN,
+        tokenPrices.token0Price,
+        token0Address,
+        provider,
+      );
+    }
+
+    if (tokenBalance1BN.gt(0)) {
+      token1ValueUSD = await getTokenUSDValue(
+        tokenBalance1BN,
+        tokenPrices.token1Price,
+        token1Address,
+        provider,
+      );
+    }
+
+    // calculate the total liquidity value of the pool at current block
+    // @dev for sanity check, liquidity should be close to the one here for usdc/eth pool https://etherscan.io/address/0x8ad599c3a0ff1de082011efddc58f1908eb6e6d8
+    /* eslint no-unused-expressions: ["error", {"allowTernary": true }] */
+    if (token0ValueUSD !== undefined && token1ValueUSD !== undefined) {
+      counter === 0
+        ? (previousLiquidity = token0ValueUSD.plus(token1ValueUSD))
+        : (currentLiquidity = token0ValueUSD.plus(token1ValueUSD));
+    } else if (token0ValueUSD !== undefined && poolEthBalanceBN.gt(0)) {
+      counter === 0
+        ? (previousLiquidity = token0ValueUSD.plus(poolEthBalanceBN))
+        : (currentLiquidity = token0ValueUSD.plus(poolEthBalanceBN));
+    } else if (token1ValueUSD !== undefined && poolEthBalanceBN.gt(0)) {
+      counter === 0
+        ? (previousLiquidity = token1ValueUSD.plus(poolEthBalanceBN))
+        : (currentLiquidity = token1ValueUSD.plus(poolEthBalanceBN));
```

Comment:
I'm confused by these two `else if` blocks, is `poolEthBalanceBN` supposed to be a backup in case you are unable to complete the token value calculation for one (but not both) of the LP tokens? Also wouldn't you need to calculate something like `poolEthValueUSD` and use that instead of just the eth balance?

Reply:
The reason `poolEthBalanceBn` is in there is because if the liquidity pool uses something like `USCD/ETH` instead of `USCD/WETH`. I was trying to account for the case where the liquidity pool swapped between an erc20 token and eth, instead of just 2 erc20 tokens. But I might have misunderstood the uniswapv3 liquidity pools and seems like the swap is _always_ between 2 erc20 tokens.

In this case, it seems like it would make the most sense to remove `poolEthBalanceBn`